ipc/provider/src/checkpoint.rs

```diff
@@ -0,0 +1,143 @@
+        })
+    }
+}
+
+impl BottomUpCheckpointManager<EthSubnetManager, EthSubnetManager> {
+    pub async fn new_evm_manager(
+        parent: Subnet,
+        child: Subnet,
+        keystore: Arc<RwLock<PersistentKeyStore<EthKeyAddress>>>,
+    ) -> Result<Self> {
+        let parent_handler =
+            EthSubnetManager::from_subnet_with_wallet_store(&parent, keystore.clone())?;
+        let child_handler = EthSubnetManager::from_subnet_with_wallet_store(&child, keystore)?;
+        Self::new(parent, child, parent_handler, child_handler).await
+    }
+}
+
+impl<P: BottomUpCheckpointRelayer, C: BottomUpCheckpointRelayer> Display
+    for BottomUpCheckpointManager<P, C>
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "bottom-up, parent: {:}, child: {:}",
+            self.metadata.parent.id, self.metadata.child.id
+        )
+    }
+}
+
+impl<
+        P: BottomUpCheckpointRelayer + Send + Sync + 'static,
+        C: BottomUpCheckpointRelayer + Send + Sync + 'static,
+    > BottomUpCheckpointManager<P, C>
+{
+    /// Getter for the parent subnet this checkpoint manager is handling
+    pub fn parent_subnet(&self) -> &Subnet {
+        &self.metadata.parent
+    }
+
+    /// Getter for the target subnet this checkpoint manager is handling
+    pub fn child_subnet(&self) -> &Subnet {
+        &self.metadata.child
+    }
+
+    /// The checkpoint period that the current manager is submitting upon
+    pub fn checkpoint_period(&self) -> ChainEpoch {
+        self.metadata.period
+    }
+
+    /// Run the bottom up checkpoint submission daemon in the background
+    pub fn run(self, validator: Address, submission_interval: Duration) {
+        tokio::spawn(async move {
+            loop {
+                if let Err(e) = self.submit_checkpoint(&validator).await {
+                    log::error!("cannot submit checkpoint for validator: {validator} due to {e}");
+                }
+
+                tokio::time::sleep(submission_interval).await;
+            }
+        });
+    }
+
+    /// Submit the checkpoint from the target validator address
+    pub async fn submit_checkpoint(&self, validator: &Address) -> Result<()> {
+        let next_submission_height = self.next_submission_height().await?;
+        let current_height = self.child_handler.current_epoch().await?;
+
+        if current_height < next_submission_height {
+            return Ok(());
+        }
+
+        let bundle = self
+            .child_handler
+            .checkpoint_bundle_at(next_submission_height)
+            .await?;
+        log::debug!("bottom up bundle: {bundle:?}");
+
+        self.parent_handler
+            .submit_checkpoint(validator, bundle)
+            .await
+            .map_err(|e| anyhow!("cannot submit bottom up checkpoint due to: {e:}"))?;
+
+        Ok(())
+    }
+
+    async fn next_submission_height(&self) -> Result<ChainEpoch> {
+        let last_checkpoint_epoch = self
```

Comment:
One caveat here, relayers are allowed to submit a checkpoint for the `lastBottomUpCheckpointHeight` (and rewarded if they do so) as long as they haven't submitted it yet, and the next checkpoint hasn't been submitted yet. Before the epoch of the next checkpoint, we should also check if we can submit the one for the current one  (even if someone has sent it already): https://github.com/consensus-shipyard/ipc-solidity-actors/blob/ad80d9b29c98473042c42b8f9cae5c7453620c25/src/subnet/SubnetActorManagerFacet.sol#L75